safe-global/web-core
# [EPIC] Relaying

## What it solves

- [x] Adds relaying as an execution method. #1700 
- [x] Dashboard widget with relaying feature information. #1701 
   ~Relaying educational series~. #1704 
- [x] Create Safe #1802 
- [x] Batch transactions modal #1818 

## TODO

- [x] Separate `NEXT_PUBLIC_SAFE_GELATO_RELAY_SERVICE_URL` into production and staging.

## How this PR fixes it
- Enables the transaction execution via a sponsored relayer.
- Displays the number of remaining hourly relays.
- Smart Contract signers can only relay a fully signed transaction.
- In a relaying execution, only the `Gas limit` can be modified.
- New widget in the dashboard containing relaying information and a link to the educational series
~- New route with the relaying educational series~
- Safe creation will be relayed if all Safe signers have relay transactions available
- Relaying is also enabled in the Batch transactions modal

## How to test it
_Execution flow_
1. In Goerli, get to the review screen of a transaction that can be executed by the connected account.
2. If the "Execute transaction" checkbox is ticked, you are able to see a menu to chose the _execution method_
3. On submitting the transaction "Via relayer" the transaction will move through the stages RELAYING -> INDEXING -> SUCCESS in the happy case.

~_Educational series_~
~1. From the dashboard, click in "Read about trial" in the new **Gasless transactions** widget.~
~2. You'll be able to navigate the relaying educational series.~

_Batch transactions_
1. Fully sign 2+ transactions so that you can batch them
2. The "How to pay" info should be present in the Modal
3. Click "Send" will relay the Multisend transaction

_Safe creation_
1. Enter the create Safe flow
2. In the review step, if all the Safe owners have available relaying transaction, a component with that info is displayed
3. On clicking "Next", the Safe creation will be relayed.
4. The transaction will be monitored until being indexed.

## Screenshots
<img width="652" alt="Screenshot 2023-03-31 at 11 28 39" src="https://user-images.githubusercontent.com/32431609/229082478-deb9fe24-bdf1-4234-bc32-b6be6b33bb4d.png">
<img width="846" alt="Screenshot 2023-03-31 at 11 29 06" src="https://user-images.githubusercontent.com/32431609/229082555-747663b2-78b3-40a8-9f53-26ae4f38b3d1.png">
<img width="820" alt="Screenshot 2023-03-31 at 11 29 19" src="https://user-images.githubusercontent.com/32431609/229082602-c4919419-a230-4aa7-aca1-526f05af362a.png">

## Checklist
